torchdata/datapipes/iter/util/rar_archive_reader.py

```diff
@@ -0,0 +1,45 @@
+import io
+import os.path
+from typing import Tuple, Iterator
+
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.iter import IterDataPipe
+from torchdata.datapipes.utils.common import validate_pathname_binary_tuple
+
+
+@functional_datapipe("read_from_rar")
+class RarArchiveReaderIterDataPipe(IterDataPipe[Tuple[str, io.BufferedIOBase]]):
+    def __init__(self, datapipe: IterDataPipe[Tuple[str, io.BufferedIOBase]]):
+        self._rarfile = self._verify_dependencies()
+        super().__init__()
+        self.datapipe = datapipe
+
+    @staticmethod
+    def _verify_dependencies():
+        try:
+            import rarfile
+        except ImportError as error:
+            raise ModuleNotFoundError(
+                "Package `rarfile` is required to be installed to use this datapipe. "
+                "Please use `pip install rarfile` or `conda -c conda-forge install rarfile` to install it."
+            ) from error
```

Comment:
I prefer mine. In your version `mypy` will probably complain everywhere, because statically `rarfile` now could be a module or `None` and you can't call anything on `None`. The perf difference is also insignificant: after the first `import`, importing again is basically a dictionary lookup with is O(1).